ot/mpz-ot-core/src/chou_orlandi/receiver.rs

```diff
@@ -0,0 +1,277 @@
+            counter,
+            choice_log,
+            decryption_keys: cached_decryption_keys,
+            ..
+        } = &mut self.state;
+
+        let private_keys = choices
+            .iter_lsb0()
+            .map(|_| Scalar::random(rng))
+            .collect::<Vec<_>>();
+
+        let (blinded_choices, decryption_keys) =
+            compute_decryption_keys(sender_base_table, &private_keys, choices, *counter);
+
+        *counter += blinded_choices.len();
+        cached_decryption_keys.extend(decryption_keys);
+
+        // If configured, log the choices
+        if self.config.receiver_commit() {
+            choice_log.extend(Vec::<u8>::from_lsb0_iter(choices.iter_lsb0()));
+        }
+
+        ReceiverPayload { blinded_choices }
+    }
+
+    /// Receives the encrypted payload from the Sender, returning the plaintext messages corresponding
+    /// to the Receiver's choices.
+    ///
+    /// # Arguments
+    ///
+    /// * `payload` - The encrypted payload from the Sender
+    pub fn receive(&mut self, payload: SenderPayload) -> Result<Vec<Block>, ReceiverError> {
+        let state::Setup {
+            decryption_keys, ..
+        } = &mut self.state;
+
+        let SenderPayload { payload } = payload;
+
+        // Check that the number of ciphertexts does not exceed the number of pending keys
+        if payload.len() > decryption_keys.len() {
+            return Err(ReceiverError::CountMismatch(
+                decryption_keys.len(),
+                payload.len(),
+            ));
+        }
+
+        // Drain the decryption keys and decrypt the ciphertexts
+        Ok(decryption_keys
+            .drain(..payload.len())
+            .zip(payload)
+            .map(
+                |((c, key), [ct0, ct1])| {
+                    if c {
+                        key ^ ct1
+                    } else {
+                        key ^ ct0
+                    }
+                },
+            )
+            .collect::<Vec<Block>>())
+    }
+
+    /// Reveals the receiver's choices to the Sender
+    pub fn reveal_choices(self) -> Result<ReceiverReveal, ReceiverError> {
+        let state::Setup {
+            choice_log: choices,
+            ..
+        } = self.state;
+
+        Ok(ReceiverReveal { choices })
+    }
+}
+
+/// Computes the blinded choices `B` and the decryption keys for the OT receiver.
+///
+/// # Arguments
+///
+/// * `base_table` - A Ristretto basepoint table from the sender's public key
+/// * `receiver_private_keys` - The private keys of the OT receiver
+/// * `choices` - The choices of the OT receiver
+/// * `offset` - The number of OTs that have already been performed
+///              (used for the key derivation tweak)
+fn compute_decryption_keys<T: BitIterable + Sync>(
+    base_table: &RistrettoBasepointTable,
+    receiver_private_keys: &[Scalar],
+    choices: &[T],
+    offset: usize,
+) -> (Vec<RistrettoPoint>, Vec<(bool, Block)>) {
+    let zero = &Scalar::ZERO * base_table;
```

Comment:
Generally, yes, but when security is on the line, it doesn't hurt to be extra safe @sinui0

Reply:
@sinui0 , we also test that we correctly understand how multiplication by 0 and by 1 works for EC points.

This is not schoolbook math, so maybe * 0 does something different than what we think it does. Having those tests in place will also alleviate the concern of whoever is auditing the code.

**Generally** we trust our dependencies. But there is nothing wrong in being extra careful and testing some corner cases in case if our deps misbehave.

I actually ran into an issue with `rs-merkle` (https://github.com/antouhou/rs-merkle/pull/30) where the lib had a bug and did not sort the leaves. So, it is useful to test dependencies when in doubt, imo.

Reply:
@sinui0 , thanks, yes, I feel strongly about it :) Let's put them under receiver.rs and later we can move them into the `point` module when we have one.